File: go/day05/main.go
package main

import (
	"fmt"
	"sort"
	"strconv"
	"strings"

	utils "github.com/MaraudingAvenger/Advent2025/common"
)

type interval struct {
	start uint64
	end   uint64
}

func parseRange(line string) (interval, bool) {
	left, right, ok := strings.Cut(strings.TrimSpace(line), "-")
	if !ok {
		panic("invalid range format: " + line)
	}
	start, err1 := strconv.ParseUint(left, 10, 64)
	end, err2 := strconv.ParseUint(right, 10, 64)
	if err1 != nil || err2 != nil {
		return interval{}, false
	}
	return interval{start: start, end: end}, true
}

func main() {
	lines, err := utils.ReadInputFile(utils.FindInputFile("inputs/day5.txt"))
	if err != nil {
		fmt.Println("Error reading input file:", err)
		return
	}

	var ranges []interval
	var ingredients []uint64

	parsingRanges := true

	for _, line := range strings.Split(lines, "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			parsingRanges = false
			continue
		}

		if parsingRanges {
			r, ok := parseRange(line)
			if ok {
				ranges = append(ranges, r)
			}
		} else {
			if value, err := strconv.ParseUint(line, 10, 64); err == nil {
				ingredients = append(ingredients, value)
			}
		}
	}

	// Sort ranges and ingredients in place.
	sort.Slice(ranges, func(i, j int) bool {
		return ranges[i].start < ranges[j].start
	})
	sort.Slice(ingredients, func(i, j int) bool {
		return ingredients[i] < ingredients[j]
	})

	// Merge ranges in place.
	mergedCount := 0
	for _, r := range ranges {
		if mergedCount == 0 {
			ranges[0] = r
			mergedCount = 1
			continue
		}

		last := &ranges[mergedCount-1]
		if r.start <= last.end+1 {
			if r.end > last.end {
				last.end = r.end
			}
		} else {
			ranges[mergedCount] = r
			mergedCount++
		}
	}
	mergedRanges := ranges[:mergedCount]

	// Count ingredients in any merged range with a linear two-pointer scan.
	count := 0
	rangeIdx := 0
	for _, ingredient := range ingredients {
		for rangeIdx < len(mergedRanges) && ingredient > mergedRanges[rangeIdx].end {
			rangeIdx++
		}
		if rangeIdx == len(mergedRanges) {
			break
		}
		if ingredient >= mergedRanges[rangeIdx].start {
			count++
		}
	}

	fmt.Printf("Part 1: %d\n", count)

	var total uint64 = 0
	for _, r := range mergedRanges {
		total += r.end - r.start + 1
	}

	fmt.Printf("Part 2: %d\n", total)
}

File: rust/day05/src/main.rs
use common::read_lines;
use common::find_input_file;

fn main() {
    let lines = read_lines(&find_input_file("inputs/day5.txt"));
    let mut ranges = Vec::new();
    let mut ingredients = Vec::new();

    let mut parsing_ranges = true;
    for line in lines {
        if line.is_empty() {
            parsing_ranges = false;
            continue;
        }

        if parsing_ranges {
            let parts = line
                .trim()
                .split("-")
                .map(|s| s.parse::<usize>().expect("Could not parse range"))
                .collect::<Vec<usize>>();
            ranges.push((parts[0], parts[1]));
        }
        else {
            ingredients.push(line.trim().parse::<usize>().expect("Could not parse ingredient"));
        }
    }

    // combine overlapping ranges
    ranges.sort_by(|a, b| a.0.cmp(&b.0));
    let mut combined_ranges = Vec::new();
    let mut current_range = ranges[0];
    for &range in &ranges[1..] {
        if range.0 <= current_range.1 {
            current_range.1 = current_range.1.max(range.1);
        } else {
            combined_ranges.push(current_range);
            current_range = range;
        }
    }
    combined_ranges.push(current_range);
    ranges = combined_ranges;
    ingredients.sort();

    let mut count = 0;
    for ingredient in ingredients {
        for &(start, end) in &ranges {
            if ingredient >= start && ingredient <= end {
                count += 1;
                break;
            }
        }
    }
    println!("Part 1: {}", count);

    let mut total_fresh: usize = 0;
    for &(start, end) in &ranges {
        total_fresh += end - start + 1;
    }

    println!("Part 2: {}", total_fresh);
}
